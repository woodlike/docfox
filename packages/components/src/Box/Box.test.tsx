import { matchers } from 'jest-emotion';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'theme-ui';
import * as React from 'react';

import { Box } from '.';
import { theme } from '..';

expect.extend(matchers);

describe('<Box />', () => {
  const id = 'Test Box';
  const { borderStyles, borderWidths, space } = theme;
  describe('Padding handling', () => {
    it('(shorthand): handles a Theme-UI shorthand padding area', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2}>{id}</Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'padding',
        `${space[2]}px ${space[2]}px ${space[2]}px ${space[2]}px`,
      );
      unmount();
    });

    it('(vertical | horizontal): handles a Theme-UI tuple padding area', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={[3, 5]}>{id}</Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'padding',
        `${space[3]}px ${space[5]}px ${space[3]}px ${space[5]}px`,
      );
      unmount();
    });

    it('(top | horizontal | bottom): handles a Theme-UI triple padding area', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={[3, 5, 9]}>{id}</Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'padding',
        `${space[3]}px ${space[5]}px ${space[9]}px ${space[5]}px`,
      );
      unmount();
    });

    it('(top | right | bottom | left): handles a Theme-UI triple padding area', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={[2, 6, 3, 2]}>{id}</Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'padding',
        `${space[2]}px ${space[6]}px ${space[3]}px ${space[2]}px`,
      );
      unmount();
    });
  });

  describe('Border width handling', () => {
    it('(shorthand): handles a Theme-UI shorthand border width', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderWidths={3}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'border-width',
        `${borderWidths[3]}px ${borderWidths[3]}px ${borderWidths[3]}px ${borderWidths[3]}px`,
      );
      unmount();
    });

    it('(vertical | horizontal): handles a Theme-UI tuple border width', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderWidths={[3, 5]}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'border-width',
        `${borderWidths[3]}px ${borderWidths[5]}px ${borderWidths[3]}px ${borderWidths[5]}px`,
      );
      unmount();
    });

    it('(top | horizontal | bottom): handles a Theme-UI triple border width', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderWidths={[3, 5, 9]}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'border-width',
        `${borderWidths[3]}px ${borderWidths[5]}px ${borderWidths[9]}px ${borderWidths[5]}px`,
      );
      unmount();
    });

    it('(top | right | bottom | left): handles a Theme-UI triple border width', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderWidths={[2, 6, 3, 2]}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'border-width',
        `${borderWidths[2]}px ${borderWidths[6]}px ${borderWidths[3]}px ${borderWidths[2]}px`,
      );
      unmount();
    });
  });

  describe('Border styles handling', () => {
    it('(shorthand): handles a Theme-UI shorthand border styles', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderStyles={3}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'border-style',
        `${borderStyles[3]}px ${borderStyles[3]}px ${borderStyles[3]}px ${borderStyles[3]}px`,
      );
      unmount();
    });

    it('(vertical | horizontal): handles a Theme-UI tuple border styles', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderStyles={[3, 5]}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'border-style',
        `${borderStyles[3]}px ${borderStyles[5]}px ${borderStyles[3]}px ${borderStyles[5]}px`,
      );
      unmount();
    });

    it('(top | horizontal | bottom): handles a Theme-UI triple border styles', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderStyles={[3, 5, 9]}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'border-style',
        `${borderStyles[3]}px ${borderStyles[5]}px ${borderStyles[9]}px ${borderStyles[5]}px`,
      );
      unmount();
    });

    it('(top | right | bottom | left): handles a Theme-UI triple border styles', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderStyles={[2, 6, 3, 2]}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toHaveStyleRule(
        'border-style',
        `${borderStyles[2]}px ${borderStyles[6]}px ${borderStyles[3]}px ${borderStyles[2]}px`,
      );
      unmount();
    });
  });

  describe('Border color handling', () => {
    it('(shorthand): handles a Theme-UI shorthand border color', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderColors={'primary'}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toMatchSnapshot();
      unmount();
    });

    it('(shorthand): handles a Theme-UI shorthand border color scale', () => {
      const borderColors = { color: 'corals', idx: 1 };
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderColors={borderColors}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toMatchSnapshot();
      unmount();
    });

    it('(vertical | horizontal):  handles a Theme-UI tuple border color', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderColors={['primary', 'secondary']}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toMatchSnapshot();
      unmount();
    });

    it('(vertical | horizontal):  handles a Theme-UI tuple border color scale', () => {
      const borderColors = [{ color: 'corals', idx: 1 }, 'secondary'];
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderColors={borderColors}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toMatchSnapshot();
      unmount();
    });

    it('(top | horizontal | bottom):  handles a Theme-UI triple border color scale', () => {
      const borderColors = [
        { color: 'whites', idx: 2 },
        { color: 'blacks', idx: 1 },
        'secondary',
      ];
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderColors={borderColors}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toMatchSnapshot();
      unmount();
    });

    it('(top | right | bottom | left):  handles a Theme-UI triple border color scale', () => {
      const borderColors = [
        { color: 'whites', idx: 2 },
        { color: 'blacks', idx: 1 },
        { color: 'grays', idx: 0 },
        { color: 'corals', idx: 1 },
      ];

      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} borderColors={borderColors}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toMatchSnapshot();
      unmount();
    });
  });

  describe('Background color handling', () => {
    it('handles a Theme-UI background color', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} backgroundColor={'primary'}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toMatchSnapshot();
      unmount();
    });

    it('handles a Theme-UI background color scale', () => {
      const backgroundColor = { color: 'grays', idx: 3 };
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box padding={2} backgroundColor={backgroundColor}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box).toMatchSnapshot();
      unmount();
    });
  });

  describe('HTML tag rendering', () => {
    it('should render a DIV tag', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box as="div" padding={2}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box.tagName).toBe('DIV');
      unmount();
    });

    it('should render a ASIDE tag', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box as="aside" padding={2}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box.tagName).toBe('ASIDE');
      unmount();
    });

    it('should render a SECTION tag', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box as="section" padding={2}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box.tagName).toBe('SECTION');
      unmount();
    });

    it('should render a MAIN tag', () => {
      const { getByText, unmount } = render(
        <ThemeProvider theme={theme}>
          <Box as="main" padding={2}>
            {id}
          </Box>
        </ThemeProvider>,
      );
      const box = getByText(id);
      expect(box.tagName).toBe('MAIN');
      unmount();
    });
  });
});
