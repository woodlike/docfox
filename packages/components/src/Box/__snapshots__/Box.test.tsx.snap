// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> Background color handling handles a Theme-UI background color 1`] = `
.emotion-0 {
  background-color: var(--theme-ui-colors-primary,rgb(255,113,99));
  padding: 6px 6px 6px 6px;
}

<div
  class="emotion-0"
>
  Test Box
</div>
`;

exports[`<Box /> Background color handling handles a Theme-UI background color scale 1`] = `
.emotion-0 {
  background-color: var(--theme-ui-colors-grays-3,rgb(89,103,113));
  padding: 6px 6px 6px 6px;
}

<div
  class="emotion-0"
>
  Test Box
</div>
`;

exports[`<Box /> Border color handling (shorthand): handles a Theme-UI shorthand border color 1`] = `
.emotion-0 {
  border-color: var(--theme-ui-colors-primary,rgb(255,113,99));
  padding: 6px 6px 6px 6px;
}

<div
  class="emotion-0"
>
  Test Box
</div>
`;

exports[`<Box /> Border color handling (shorthand): handles a Theme-UI shorthand border color scale 1`] = `
.emotion-0 {
  border-color: var(--theme-ui-colors-corals-1,rgb(229,85,78));
  padding: 6px 6px 6px 6px;
}

<div
  class="emotion-0"
>
  Test Box
</div>
`;

exports[`<Box /> Border color handling (top | horizontal | bottom):  handles a Theme-UI triple border color scale 1`] = `
.emotion-0 {
  border-color: var(--theme-ui-colors-whites-2,rgb(240,240,240)) var(--theme-ui-colors-blacks-1,rgb(51,51,51)) var(--theme-ui-colors-secondary,rgb(229,85,78)) var(--theme-ui-colors-blacks-1,rgb(51,51,51));
  padding: 6px 6px 6px 6px;
}

<div
  class="emotion-0"
>
  Test Box
</div>
`;

exports[`<Box /> Border color handling (top | right | bottom | left):  handles a Theme-UI triple border color scale 1`] = `
.emotion-0 {
  border-color: var(--theme-ui-colors-whites-2,rgb(240,240,240)) var(--theme-ui-colors-blacks-1,rgb(51,51,51)) var(--theme-ui-colors-grays-0,rgb(222,223,224)) var(--theme-ui-colors-corals-1,rgb(229,85,78));
  padding: 6px 6px 6px 6px;
}

<div
  class="emotion-0"
>
  Test Box
</div>
`;

exports[`<Box /> Border color handling (vertical | horizontal):  handles a Theme-UI tuple border color 1`] = `
.emotion-0 {
  border-color: var(--theme-ui-colors-primary,rgb(255,113,99)) var(--theme-ui-colors-secondary,rgb(229,85,78)) var(--theme-ui-colors-primary,rgb(255,113,99)) var(--theme-ui-colors-secondary,rgb(229,85,78));
  padding: 6px 6px 6px 6px;
}

<div
  class="emotion-0"
>
  Test Box
</div>
`;

exports[`<Box /> Border color handling (vertical | horizontal):  handles a Theme-UI tuple border color scale 1`] = `
.emotion-0 {
  border-color: var(--theme-ui-colors-corals-1,rgb(229,85,78)) var(--theme-ui-colors-secondary,rgb(229,85,78)) var(--theme-ui-colors-corals-1,rgb(229,85,78)) var(--theme-ui-colors-secondary,rgb(229,85,78));
  padding: 6px 6px 6px 6px;
}

<div
  class="emotion-0"
>
  Test Box
</div>
`;
